src/fury/encoder/preprocessor.h

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2023 The Fury Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
```

Comment:
Do we need to add some comments for this preprocessor? such as:
// Use a two-step macro expansion process, i.e. "macro indirection"  to ensure that arguments passed to the macros are fully expanded before they are concatenated.

Reply:
Added.